torch/csrc/DeviceAccelerator.cpp

```diff
@@ -0,0 +1,104 @@
+#include <c10/core/DeviceGuard.h>
+#include <torch/csrc/DeviceAccelerator.h>
+#include <torch/csrc/utils/device_lazy_init.h>
+
+namespace torch::accelerator {
+
+void initModule(PyObject* module) {
```

Comment:
The reason that we introduce `current_accelerator` is that perhaps the user needs to know what the current accelerator is to write some device-specific code.
`current_accelerator` is based on `_accelerator_getAccelerator` which is a C++ python binding function.
The other APIs leverage the `at::getAccelerator` and `VirtualGuardImpl` registration mechanisms to implement their functionality. The call stack is like this:
`torch.accelerator.set_device`(python) -> `_accelerator_setDevice`(C++) -> `at::getAccelerator` & `VirtualGuardImpl.setDevice`

Reply:
In other words, do you mean whether it is possible to implement those other APIs by Python only?  I said yes. The only method I could think is like
```python
def set_device(device):
    current_accelerator = torch.accelerator.current_accelerator()
    # call to device module `torch.cuda` or `torch.xpu`
    get_device_module(current_accelerator).set_device()
```
But I don't think this is a good implementation. The reason is:
1. `torch.cuda.set_device` these device-specific runtime APIs should be deprecated in the future, And recommended to use `torch.accelerator.set_device` instead.
2. `get_device_module(current_accelerator).set_device()` has an extra python call overhead.

Why do we choose to reuse `VirtualGuardImpl` registration mechanisms?
In my future plan, I propose to unify these C++ APIs to the at::accelerator namespace and implement these APIs via reusing the `DeviceGuardImpl` registration mechanism as well. They should be `at::accelerator::current_stream()`, `at::accelerator::current_device()` etc in my image.
My all picture is that Stream and Device APIs leverage `DeviceGuardImpl` registration mechanism. Allocator APIs leverage Allocator registration mechanism. And Generator APIs should have its own registration mechanism, similar to Allocator. This way, we can unify all runtime APIs in both C++ and Python side while C++ APIs might be under namespace at::accelerator and Python APIs might be under namespace `torch.accelerator`. The code example like this:
```C++
namespace at::accelerator {
void setDevice(DeviceIndex device_index) {
  const auto device_type = at::getAccelerator(true).value();
  // If device index is negative, no-op
  if (device_index < 0) {
    return;
  }
  c10::impl::VirtualGuardImpl impl(device_type);
  impl.setDevice({device_type, device_index});
}
}
```
And then we declare the python C++ bind function, like this
```C++
m.def("_accelerator_setDevice", [](c10::DeviceIndex device_index) {
  const auto device_type = at::getAccelerator(true).value();
  torch::utils::maybe_initialize_device(device_type);
  at::accelerator::setDevice(device_index);
}
```
The final python frontend code, like this
```python
# under `torch.accelerator` namespace
def set_device(device: _device_t) -> None:
    device_index = _get_device_index(device)
    torch._C._accelerator_setDevice(device_index)
```
Considering the above design, we can unify all these device-agnostic APIs both in C++ and Python side.
Call `at::accelerator::foo` in C++ and `torch.accelerator.foo` in Python.

The reason for not adding all API interface to DeviceHooksInterface is that
1. It can prevent the content under the DeviceHooksInterface from becoming too extensive.
2. allows each backend to register only what they need.
3. in my understanding, `HooksInterface` is used to solve the device-specific issue as a safe bridge between ATen libaray and libtorch libaray, like `versionCuDNN`, `versionCUDART` for cudahooks. And `getGlobalIdxFromDevice` for xpuhooks. They are all device-specific.

Reply:
Yes, I will unify these C++ APIs in the following PR.